server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -0,0 +1,310 @@
+                if (newSnapshotShardSizes != null) {
+                    knownSnapshotShardSizes = newSnapshotShardSizes.build();
+                }
+            }
+
+            final int nbFetchers = Math.min(unknownShards, maxConcurrentFetches);
+            for (int i = 0; i < nbFetchers; i++) {
+                final int activeFetches = concurrentFetches.incrementAndGet();
+                if (activeFetches < maxConcurrentFetches + 1) {
+                    fetchNextSnapshotShard();
+                } else {
+                    final int value = concurrentFetches.decrementAndGet();
+                    assert value >= 0 : "Unexpected value: " + value;
+                }
+            }
+        } else {
+            synchronized (this) {
+                // information only needed on current master
+                knownSnapshotShardSizes = ImmutableOpenMap.of();
+                isMaster = false;
+            }
+        }
+    }
+
+    private void fetchNextSnapshotShard() {
+        boolean success = false;
+        try {
+            final SnapshotShard snapshotShard = queue.poll(0L, TimeUnit.MILLISECONDS);
+            if (snapshotShard != null) {
+                threadPool.executor(executorName()).execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() {
+                        if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                            logger.debug("skipping snapshot shard size retrieval for {} as node is no longer master", snapshotShard);
+                            return;
+                        }
+
+                        final RepositoriesService repositories = repositoriesService.get();
+                        assert repositories != null;
+                        final Repository repository = repositories.repository(snapshotShard.snapshot.getRepository());
+
+                        logger.debug("fetching snapshot shard size for {}", snapshotShard);
+                        final IndexShardSnapshotStatus status = repository.getShardSnapshotStatus(
+                            snapshotShard.snapshot().getSnapshotId(),
+                            snapshotShard.index(),
+                            snapshotShard.shardId()
+                        );
+
+                        final long snapshotShardSize = status.asCopy().getTotalSize();
+                        logger.debug("snapshot shard size for {}: {} bytes", snapshotShard, snapshotShardSize);
+
+                        boolean updated = false;
+                        synchronized (InternalSnapshotsInfoService.this) {
+                            if (isMaster) {
+                                final ImmutableOpenMap.Builder<SnapshotShard, Long> newSnapshotShardSizes =
+                                    ImmutableOpenMap.builder(knownSnapshotShardSizes);
+                                updated = newSnapshotShardSizes.put(snapshotShard, snapshotShardSize) == null;
```

Comment:
Isn't `updated` always `true` since we won't double schedule a job for a certain shard's size ever?

Reply:
Indeed - I added an assertion on this in 2faf52519a6d663559eacac5592a7cbe6a317f80